libs/angular/src/modules/vault-filter/models/vault-filter.model.spec.ts

```
@@ -0,0 +1,205 @@
+import { CipherType } from "@bitwarden/common/enums/cipherType";
+import { CipherView } from "@bitwarden/common/models/view/cipherView";
+
+import { VaultFilter } from "./vault-filter.model";
+
+describe("VaultFilter", () => {
+  describe("filterFunction", () => {
+    it("should return true when filter is set to all statuses", () => {
+      const cipher = createCipher();
+      const filterFunction = createFilterFunction({ status: "all" });
+
+      const result = filterFunction(cipher);
+
+      expect(result).toBe(true);
+    });
+
+    it("should return true when filter is set to favorites and cipher is favorite", () => {
```

There's a lot of repetition in the test names and it's difficult to identify related tests. I suggest nesting them again somehow, like:

```ts
describe("filterFunction", () => {
  describe("filters favorites", () => {
    it("when the cipher is a favorite", () => ...);
    it("when the cipher is not a favourite", () => ...);
  });
  describe("filters deleted items", () => {
    it("when the cipher is deleted", () => ...);
```

and so on.

I don't even think you need to state "returns true" etc, if it's otherwise clear what you're testing. Maybe you can think of better wording, but you get the idea.

---

Ah, thanks for pointing that out, grouping sounds like a great idea!

Regarding the test naming, I am using the `ShouldWhen` naming convention as stated in the [TDD Manifesto](https://tddmanifesto.com/a-clean-test/). Maybe one could rephrase the name, but I'd still like to keep the convention

> ShouldWhen
> 
> ShouldWhen template is an easy to read and widely used naming template. The name first reveals the expected behavior, then the precondition of the given test case.
> 
> Examples:
>
>    - ShouldHaveUserLoggedIn_whenUserLogsIn
>    - ShouldReturnCreditCardIsInvalidValidationError_whenTheOrderIsPurchasedWithInvalidCreditCardData
>    - ShouldIncreasebalanceWith100Euro_when100EuroDepositIsMadeToTheAccount

---

Ah, test naming, we could bikeshed on that all day 😁 

I am not particularly a fan of the ShouldWhen format, to me `ShouldReturnCreditCardIsInvalidValidationError_whenTheOrderIsPurchasedWithInvalidCreditCardData` is an example of a fairly unreadable test name.

I like a plain English approach, especially here (in Jest) where we get to use strings instead of function names. I also think that "includes/excludes" explains more than "should return true/false" - you can always check the assert statement if you want to know the expected return value, I think it's more useful that the test name describes the feature/functionality being tested.

All that said - we have no style guide here, so you can keep the current format. 

My only strong recommendation would be to add "given a" to the inner `describe` blocks. e.g. "given a generic cipher", "given a favorite cipher", "given a deleted cipher", etc... because you're describing the argument it's being given rather than, say, the system under test. I think it makes the context a lot clearer and it flows better (the describe + it blocks should form a rough sentence, like "filter function given a generic cipher should return true when filtering for favorites").

---

> I am not particularly a fan of the ShouldWhen format, to me ShouldReturnCreditCardIsInvalidValidationError_whenTheOrderIsPurchasedWithInvalidCreditCardData is an example of a fairly unreadable test name.

I always assumed that the `camelCase_underscoreStyle` was just for an example for when you have to use the method name as the test name, meaning it would become `it('should return credit card is invalid validation error when the order is purchased with invalid credit card data')` when that is not a limitiation. But you have a very good point, bike shedding 😄

For me the _most important part is that the state under test is described in the name_. So `should throw error when email contains invalid characters` instead of `should throw error` which contains lots of assertions for all the error cases. Other than that I am pretty flexible about naming conventions 🙂 

Awesome suggestion, I really like `given a ...`! I'll add it straight away!